internal/sdkclient/sdkclient.go

```diff
@@ -1,28 +1,96 @@
 package sdkclient
 
 import (
+	"context"
 	"fmt"
 	"io"
 	"net/http"
 	"net/url"
 	"strings"
+	"time"
 
-	"github.com/hookdeck/hookdeck-go-sdk/client"
-	"github.com/hookdeck/hookdeck-go-sdk/option"
+	"golang.org/x/time/rate"
 )
 
+// Client is the main client struct exposed to the provider.
 type Client struct {
-	*client.Client
-
-	RawClient *RawClient
+	RawClient RawClientInterface
 }
 
 const (
 	defaultAPIBase = "api.hookdeck.com"
-	apiVersion     = "2025-07-01"
 )
 
-func InitHookdeckSDKClient(apiBase string, apiKey string, maxAttempts int, providerVersion string) Client {
+// RawClientInterface defines the contract for sending requests.
+type RawClientInterface interface {
+	SendRequest(ctx context.Context, method, path string, opts *RequestOptions) (*http.Response, error)
+}
+
+// RateLimiter interface allows for custom rate limiting implementations.
+type RateLimiter interface {
+	Wait(ctx context.Context) error
+}
+
+// RawClient is the concrete implementation with rate limiting.
+type RawClient struct {
+	apiKey      string
+	apiBase     string
+	header      http.Header
+	httpClient  HTTPDoer
+	rateLimiter RateLimiter
+}
+
+// HTTPDoer allows for mocking HTTP client in tests.
+type HTTPDoer interface {
+	Do(req *http.Request) (*http.Response, error)
+}
+
+// RequestOptions contains optional parameters for requests.
+type RequestOptions struct {
+	Body        io.Reader
+	Headers     http.Header
+	QueryParams url.Values
+}
+
+// RawClientOption allows configuring the RawClient.
+type RawClientOption func(*RawClient)
+
+// WithHTTPClient sets a custom HTTP client (useful for testing).
+func WithHTTPClient(client HTTPDoer) RawClientOption {
+	return func(c *RawClient) {
+		c.httpClient = client
+	}
+}
+
+// WithRateLimiter sets a custom rate limiter.
+func WithRateLimiter(limiter RateLimiter) RawClientOption {
```

Comment:
Is there a reason we took this approach rather than adhering to the 429 responses from the Hookdeck API and using the `Retry-After` header value etc. https://hookdeck.com/docs/api#rate-limits?

Reply:
I like the idea of being a bit more explicit instead of relying on the header, but also I wasn't aware of the headers either. The header is probably more efficient in the sense that we will maximize the throughput, so that could be something we can consider.

I guess I was thinking of the scenario where there are 1000 resources that need to be refreshed. If we send all 1000 requests at once, 760 will be told to send another request in about a min. Then we'll send all 760 again, then 520, etc.

With this rate limit logic, we will go a lot slower, spreading the requests over the course of a min (4 requests per second) instead of a big spike.

Do let me know if we should revise this logic a bit to maximize the rate limit pool!

Reply:
I think it's fine. If we get feedback on things being too slow we have an opportunity for some optimization.